src/compiler.rkt

```diff
@@ -219,3 +194,168 @@
 ;; Like `compile-progs`, but a single prog.
 (define (compile-prog expr ctx)
   (compose first (compile-progs (list expr) ctx)))
+
+(define (backward-pass ivec varc vregs vprecs vstart-precs)
+  (vector-fill! vprecs 0)
+  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)]
+        [n (in-range (- (vector-length vregs) 1) -1 -1)])
+
+    (define op (car instr))
+    (define tail-registers (rest instr))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define exps-from-above (vector-ref vprecs n))
+    (define final-parent-precision (min (*max-mpfr-prec*) (+ exps-from-above (vector-ref vstart-precs n))))
+    (when (equal? final-parent-precision (*max-mpfr-prec*))
+      (*sampling-iteration* (*max-sampling-iterations*)))
+    (vector-set! vprecs n final-parent-precision)
+
+    (define new-exponents 0)
+    (cond
+      [(equal? op ival-add)
+       (define xlo (ival-lo (first srcs)))
+       (define xlo-exp (true-exponent xlo))
+       (define xlo-sgn (bigfloat-signbit xlo))
+       (define ylo (ival-lo (second srcs)))
+       (define ylo-exp (true-exponent ylo))
+       (define ylo-sgn (bigfloat-signbit ylo))
+       (define outlo-exp (true-exponent (ival-lo output)))
+       ; consider only lower bound
+       (set! new-exponents
+             (max 0
+                  (match (and (not (equal? xlo-sgn ylo-sgn))
+                              (>= 1 (abs (- xlo-exp ylo-exp))))
+                    [#f (- (max xlo-exp ylo-exp) outlo-exp)]
+                    [#t (+ (get-slack) (- (max xlo-exp ylo-exp) outlo-exp))])))]
+      
+      [(equal? op ival-sub)
+       (define xlo (ival-lo (first srcs)))
+       (define xlo-exp (true-exponent xlo))
+       (define xlo-sgn (bigfloat-signbit xlo))
+       (define yhi (ival-lo (second srcs)))
+       (define yhi-exp (true-exponent yhi))
+       (define yhi-sgn (bigfloat-signbit yhi))
+       (define outlo-exp (true-exponent (ival-lo output)))
+               
+       ; Consider only lower bound
+       (set! new-exponents
+         (max 0
+              (match (and (equal? xlo-sgn yhi-sgn)
+                          (>= 1 (abs (- xlo-exp yhi-exp))))
+                [#f (- (max xlo-exp yhi-exp) outlo-exp)]
+                [#t (+ (get-slack) (- (max xlo-exp yhi-exp) outlo-exp))])))]
+      
+      [(equal? op ival-pow)
+       ; log[Гpow] = max[ log(y) , log(y) + log[log(x)] ]
+       (define xlo-exp (true-exponent (ival-lo (first srcs))))
+       (define xhi-exp (true-exponent (ival-hi (first srcs))))
+       (define ylo-exp (true-exponent (ival-lo (second srcs))))
+       (define yhi-exp (true-exponent (ival-hi (second srcs))))
+
+       (set! new-exponents
+             (if (> (max xlo-exp xhi-exp) 2) ; if x >= 4 (actually 7.3890561), then at least 1 additional bit is needed
+                 (+ (max ylo-exp yhi-exp) 30)
```

Comment:
I don't understand the comment and also the `30` here. I guess you're just saying that `log(log(x)) <= 30` and then only doing that for large `x` values, right?

Reply:
That's fine

Reply:
Yeah, that was my logic